apollo-ios-codegen/Sources/GraphQLCompiler/JavaScript/src/__tests__/deferDirectiveTests.ts

```
@@ -16,11 +16,151 @@ import {
+        }
+      }
+    }
+    `;
+
+    const document: DocumentNode = parseOperationDocument(
+      new Source(documentString, "Test Query", { line: 1, column: 1 })
+    );
+
+    it("should pass validation", () => {
+      const validationErrors: readonly GraphQLError[] = validateDocument(schema, document, emptyValidationOptions)
+
+      expect(validationErrors).toHaveLength(0)
+    });
+  })
+
+  // This test is being used as a fail-safe to automatically know when graphql-js removes the 'experimental'
+  // designation from the defer directive and it is automatically made available to operations. When this 
+  // test starts failing the automatic addition of the experimental defer directive we do in loadSchemaFromSources
+  // must be removed.
+  describe("does not add a defer directive", () => {
+    // Note we're not calling loadSchemaFromSources which adds the experimental defer directive
+    const schemaDocument = parse(schemaSDL)
+    const schema: GraphQLSchema = buildASTSchema(schemaDocument, { assumeValid: true, assumeValidSDL: true });
+
+    const documentString: string = `
+    query Test {
+      allAnimals {
+        ... on Animal @defer(label: "deferred") {
+          species
+        }
+      }
+    }
+    `;
+
+    const document: DocumentNode = parseOperationDocument(
+      new Source(documentString, "Test Query", { line: 1, column: 1 })
+    );
+
+    it("should fail validation", () => {
+      const validationErrors: readonly GraphQLError[] = validateDocument(schema, document, emptyValidationOptions)
+
+      expect(validationErrors).toHaveLength(1)
+      expect(validationErrors[0].message).toEqual(
+        "Unknown directive \"@defer\"."
+      )
+    });
+  });
+})
+
+describe("test invalid directive definitions", () => {
+
+  // Directive Definition Tests
+
+  describe("given mismatched location in directive", () => {
+    const schemaSDL: string = `
+    directive @defer(label: String, if: Boolean! = true) on FIELD
```

If the defer directive is _only_ on `FIELD` than we should throw an error, but if it's on _both_ fields and `FRAGMENT_SPREAD | INLINE_FRAGMENT` then I think we should allow it here. We should only fail if they actually try to use a `@defer` directive on a field.

---

As we spoke about earlier in the week, the behaviour here now is to print a warning and automatically replace the unsupported definition with the experimental graphql-js defined directive that we do support.